primitives/tracing/src/types.rs

```diff
@@ -0,0 +1,648 @@
+				}
+			}
+			None
+		}).collect::<Vec<_>>();
+
+		let values = match values.len() {
+			0 => metadata.fields().value_set(&[]),
+			1 => slice_to_values!(metadata.fields(), values.as_slice(), 1),
+			2 => slice_to_values!(metadata.fields(), values.as_slice(), 2),
+			3 => slice_to_values!(metadata.fields(), values.as_slice(), 3),
+			4 => slice_to_values!(metadata.fields(), values.as_slice(), 4),
+			5 => slice_to_values!(metadata.fields(), values.as_slice(), 5),
+			6 => slice_to_values!(metadata.fields(), values.as_slice(), 6),
+			7 => slice_to_values!(metadata.fields(), values.as_slice(), 7),
+			8 => slice_to_values!(metadata.fields(), values.as_slice(), 8),
+			9 => slice_to_values!(metadata.fields(), values.as_slice(), 9),
+			10 => slice_to_values!(metadata.fields(), values.as_slice(), 10),
+			11 => slice_to_values!(metadata.fields(), values.as_slice(), 11),
+			12 => slice_to_values!(metadata.fields(), values.as_slice(), 12),
+			13 => slice_to_values!(metadata.fields(), values.as_slice(), 13),
+			14 => slice_to_values!(metadata.fields(), values.as_slice(), 14),
+			15 => slice_to_values!(metadata.fields(), values.as_slice(), 15),
+			16 => slice_to_values!(metadata.fields(), values.as_slice(), 16),
+			17 => slice_to_values!(metadata.fields(), values.as_slice(), 17),
+			18 => slice_to_values!(metadata.fields(), values.as_slice(), 18),
+			19 => slice_to_values!(metadata.fields(), values.as_slice(), 19),
+			20 => slice_to_values!(metadata.fields(), values.as_slice(), 20),
+			21 => slice_to_values!(metadata.fields(), values.as_slice(), 21),
+			22 => slice_to_values!(metadata.fields(), values.as_slice(), 22),
+			23 => slice_to_values!(metadata.fields(), values.as_slice(), 23),
+			24 => slice_to_values!(metadata.fields(), values.as_slice(), 24),
+			25 => slice_to_values!(metadata.fields(), values.as_slice(), 25),
+			26 => slice_to_values!(metadata.fields(), values.as_slice(), 26),
+			27 => slice_to_values!(metadata.fields(), values.as_slice(), 27),
+			28 => slice_to_values!(metadata.fields(), values.as_slice(), 28),
+			29 => slice_to_values!(metadata.fields(), values.as_slice(), 29),
+			30 => slice_to_values!(metadata.fields(), values.as_slice(), 30),
+			31 => slice_to_values!(metadata.fields(), values.as_slice(), 31),
+			_ => slice_to_values!(metadata.fields(), &values[..32], 32),
+		};
+
+		f(metadata, &values)
+	}
+
+	fn make_hex(inp: &[u8]) -> String {
+		inp.iter().map(|c| format!("{:02x}", c)).collect::<String>()
+	}
+
+	impl WasmValue {
+
+		fn as_value<'a>(&'a mut self) -> Option<&'a dyn Value> {
+			match self {
+				// We convert to String
+				WasmValue::Formatted(i) | WasmValue::Str(i) => {
+					if let Ok(s) = String::from_utf8(i.to_vec()) {
+						*self = WasmValue::Extracted(display(s))
+					} else {
+						*self = WasmValue::Extracted(display(make_hex(&i)))
+					}
+				}
+				WasmValue::Encoded(i) => {
+					*self = WasmValue::Extracted(display(make_hex(&i)))
+				}
+				_ => {}
+			}
+
+			match self {
```

Comment:
This is one of the worst parts of it all IMHO.

The value we are passing over, in order to be able to do it, are `Vec<u8>` for debug, string and `scale`-codec-stuff (although that isn't actually supported by the latest API anymore). But for a `ValueSet` they must be converted to `Value`, which `tracing_core` prevents anyone else from implemented by having an internally private-trait `Seal` you'd also have to implement, but simply can't.

So, in order to be able to do that, we have to convert the inner-values into things that have `Value` implemented, which is only a small set of pre-defined objects – and `Vec<u8>` nor `[u8]` are among them. So, what we do here is convert these into their `DisplayValue` holding the `String` and need to replace our internal value with it, to make sure that the `&dyn Value` we have to leak (we can't provide the inner value in other ways because they are converted to the static-array later – see above) sticks around for at least the processing....